dynamic_array/include/dynamic_array/array.h

```diff
@@ -0,0 +1,303 @@
+  // value, returns |kIndexNotFound| if not found
+  int find(const T& item);
+
+  // Swap values inside |first| and |second|.
+  // Follow copy-and-swap idiom
+  // https://stackoverflow.com/questions/3279543/what-is-the-copy-and-swap-idiom
+  friend void swap<T>(Array<T>& rhs, Array<T>& lsh);
+
+ private:
+  // If |index| is a negative value or greater than
+  // |size_|, log an error and terminate program.
+  void throw_out_of_range_error_if_needed(int index);
+
+  // If |new_size| is equal or greater than |capacity_|, allocate
+  // new array with double capacity. If |new_size| is equal or less
+  // than |capacity_| / 4, allocate new array with half of capacity.
+  void reallocate_if_needed(size_t new_size);
+
+  // Deep copy |items_| from |array|
+  void deep_copy(const Array<T>& array);
+
+  // Raw array where items are stored.
+  T* items_{nullptr};
+
+  // Number of items are currently stored in array.
+  size_t size_{0};
+
+  // Represent how many total items array can store without reallocation.
+  size_t capacity_{0};
+};
+
+}  // namespace td
+
+/****************  Array implmentation ****************/
+namespace td {
+
+template <typename T>
+void swap(Array<T>& rhs, Array<T>& lhs) {
+  using std::swap;
+
+  swap(rhs.size_, lhs.size_);
+  swap(rhs.capacity_, lhs.capacity_);
+  swap(rhs.items_, lhs.items_);
+}
+
+// Public
+template <typename T>
+Array<T>::Array() : Array(min_capacity) {}
+
+template <typename T>
+Array<T>::Array(size_t capacity) : capacity_(capacity) {
+  items_ = new T[capacity_];
+}
+
+template <typename T>
+Array<T>::Array(std::initializer_list<T>&& il) : Array() {
+  for (const T& t : il) {
+    append(t);
+  }
+}
+
+template <typename T>
+Array<T>::Array(const Array<T>& other) {
+  deep_copy(other);
+}
+
+template <typename T>
+Array<T>::Array(Array<T>&& other)
+    : items_(other.items_), size_(other.size_), capacity_(other.capacity_) {
+  other.items_ = nullptr;
+}
+
+template <typename T>
+Array<T>& Array<T>::operator=(const Array<T>& other) {
+  Array<T> temp_array(other);
+  swap(*this, temp_array);
+  return *this;
+}
+
+template <typename T>
+Array<T>& Array<T>::operator=(Array<T>&& other) {
+  // Maybe shouldn't use |swap| method for move assignment operator
+  // https://stackoverflow.com/questions/6687388/why-do-some-people-use-swap-for-move-assignments
+  Array<T> temp_array = std::move(other);
+  swap(*this, temp_array);
+  return *this;
+}
+
+template <typename T>
+Array<T>::~Array() {
+  delete[] items_;
+}
+
+template <typename T>
+size_t Array<T>::size() {
+  return size_;
+}
+
+template <typename T>
+size_t Array<T>::capacity() {
+  return capacity_;
+}
+
+template <typename T>
+bool Array<T>::is_empty() {
+  return size_ == 0;
+}
+
+template <typename T>
+T Array<T>::item_at(int index) {
+  throw_out_of_range_error_if_needed(index);
+
+  return items_[index];
+}
+
+template <typename T>
+void Array<T>::append(const T& item) {
+  reallocate_if_needed(size_ + 1);
```

Comment:
It's redundant and coslty. Just need to check if `(size_ + 1) >= capacity_` then reallocate double of `capacity_`.